docarray/array/array.py

```diff
@@ -13,34 +32,299 @@ class DocumentArray(
+        ] = None
+
     def __class_getitem__(cls, item: Type[BaseDocument]):
         if not issubclass(item, BaseDocument):
             raise ValueError(
                 f'DocumentArray[item] item should be a Document not a {item} '
             )
 
-        class _DocumenArrayTyped(DocumentArray):
+        class _DocumentArrayTyped(cls):  # type: ignore
             document_type: Type[BaseDocument] = item
 
-        for field in _DocumenArrayTyped.document_type.__fields__.keys():
+        for field in _DocumentArrayTyped.document_type.__fields__.keys():
 
             def _property_generator(val: str):
-                return property(lambda self: self._get_documents_attribute(val))
+                def _getter(self):
+                    return self._get_array_attribute(val)
+
+                def _setter(self, value):
+                    self._set_array_attribute(val, value)
 
-            setattr(_DocumenArrayTyped, field, _property_generator(field))
+                # need docstring for the property
+                return property(fget=_getter, fset=_setter)
+
+            setattr(_DocumentArrayTyped, field, _property_generator(field))
             # this generates property on the fly based on the schema of the item
 
-        _DocumenArrayTyped.__name__ = f'DocumentArray[{item.__name__}]'
-        _DocumenArrayTyped.__qualname__ = f'DocumentArray[{item.__name__}]'
+        _DocumentArrayTyped.__name__ = f'DocumentArray[{item.__name__}]'
+        _DocumentArrayTyped.__qualname__ = f'DocumentArray[{item.__name__}]'
+
+        return _DocumentArrayTyped
+
+    def __getitem__(self, item):
+        if self.is_stacked():
+            doc = super().__getitem__(item)
+            # NOTE: this could be speed up by using a cache
+            for field in self._columns.keys():
+                setattr(doc, field, self._columns[field][item])
+            return doc
+        else:
+            return super().__getitem__(item)
+
+    def stack(self):
+        """
+        Calling this method will make the DocumentArray enter stacked mode. You should
+        call this method only if the DocumentArray is already in unstacked mode
+        (the default mode). (Calling it while being already in stacked mode will have
+        no effect)
+
+        When entering stack mode the DocumentArray will create a column for each field
+        of the Document that are Tensor or that are Nested Document that contains at
+        least one Tensor field. This is useful when you want to perform operations on
+        the whole array at once. In stack mode, accessing or setting fields of the
+        DocumentArray will access or set the column of the array.
+
+
+        IMPORTANT: in stacked_mode you cannot add or remove Document from the array.
+        You can only modify the fields of the DocumentArray. This is intentionally done
+        because extending or inserting a column is slow. You should rather use the
+        unstack mode for such operation.
```

Comment:
These explanations should be either here, or at the top level docstring, not both. Let's keep a single source of truth and link